linux/src/threadpool_linux.c

```diff
@@ -501,29 +531,156 @@ int threadpool_schedule_work(THREADPOOL_HANDLE threadpool, THREADPOOL_WORK_FUNCT
 
 int threadpool_timer_start(THREADPOOL_HANDLE threadpool, uint32_t start_delay_ms, uint32_t timer_period_ms, THREADPOOL_WORK_FUNCTION work_function, void* work_function_ctx, TIMER_INSTANCE_HANDLE* timer_handle)
 {
-    (void)threadpool;
-    (void)start_delay_ms;
-    (void)timer_period_ms;
-    (void)work_function;
-    (void)work_function_ctx;
-    (void)timer_handle;
-    return MU_FAILURE;
+    int result;
+
+    if (
+        threadpool == NULL ||
+        work_function == NULL ||
+        timer_handle == NULL
+        )
+    {
+        LogError("Invalid args: THREADPOOL_HANDLE threadpool = %p, uint32_t start_delay_ms = %" PRIu32 ", uint32_t timer_period_ms = %" PRIu32 ", THREADPOOL_WORK_FUNCTION work_function = %p, void* work_function_context = %p, TIMER_INSTANCE_HANDLE* timer_handle = %p",
+            threadpool, start_delay_ms, timer_period_ms, work_function, work_function_ctx, timer_handle);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        (void)interlocked_increment(&threadpool->pending_call_count);
+
+        TIMER_INSTANCE* timer_instance = malloc(sizeof(TIMER_INSTANCE));
+        if (timer_instance == NULL)
+        {
+            LogError("Failure allocating Timer Instance");
+        }
+        else
+        {
+            timer_instance->work_function = work_function;
+            timer_instance->work_function_ctx = work_function_ctx;
+
+            // Setup the timer
+            struct sigevent sigev = {0};
+            timer_t time_id = 0;
+
+            sigev.sigev_notify          = SIGEV_THREAD;
+            sigev.sigev_notify_function = on_timer_callback;
+            sigev.sigev_value.sival_ptr = timer_instance;
+
+            if (timer_create(CLOCK_REALTIME, &sigev, &time_id) != 0)
+            {
+                char err_msg[128] = {0};
+                (void)strerror_r(errno, err_msg, 128);
+                LogError("Failure calling timer_create. Error: %s", MU_P_OR_NULL(err_msg));
+            }
+            else
+            {
+                struct itimerspec its = {0};
+                its.it_value.tv_sec = start_delay_ms / 1000;
```

Comment:
>start_delay_ms / 1000; [](http://example.com/codeflow?start=38&length=22)

I keep seeing this where we divide with N and get the quotient of N too at the same time.

Can "div" (or similar from stdlib.h) be used here?